### ISSUE TITLE
Cannot reproduce visualization

### ISSUE THREAD

**Issue by yeshPP** [datalogue/keras-attention#8]
Enjoyed your blog post and thank you for the code. I am, however, having trouble reproducing your visualization of the attention layer. If I use your sample vocabulary (sample_*_vocab.json) and your sample weights (sample_NMT.49.0.01.hdf5) the visualization works as expected. However, if I generate a fresh training and validation set, train for ~50 epochs (acc=0.9952), and generate the visualization plots with those weights it makes little sense. Unlike the pristine plots on your blog post, these plots seem to indicate that there are just 2 or 3 points of interest in the attention map. The predicted text however is correct every time.  Here's  an example - http://bit.ly/2umNcvv 

Is the problem that the attention layer takes significantly longer to train, or is it something that I am missing? 

Also, during training all_acc and val_all_acc are always zero. Any insights? 

I am running TF-1.1.0 and Keras-2.0.6.

**Comment by zafarali** [comment]
Hi thanks for trying it out!

I'll launch an experiment today to see if I can reproduce it. In the
meantime can you give me the command you used to run `visualize.py`?

**Comment by yeshPP** [comment]
Thanks for testing it out. I typically run visualize as follows:
python visualize.py -e "june 8, 2015" -w weights/NMT.07-0.02.hdf5

where NMT.07-0.02.hdf5 are my weights after 50 epochs. I also changed the the names of the sample vocabulary in visualize.py as follows:

```
diff --git a/visualize.py b/visualize.py
index 8105681..6f4741f 100644
--- a/visualize.py
+++ b/visualize.py
@@ -19,8 +19,10 @@ def load_examples(file_name):
 if not os.path.exists(os.path.join(HERE, 'attention_maps')):
     os.makedirs(os.path.join(HERE, 'attention_maps'))

-SAMPLE_HUMAN_VOCAB = os.path.join(HERE, 'data', 'sample_human_vocab.json')
-SAMPLE_MACHINE_VOCAB = os.path.join(HERE, 'data', 'sample_machine_vocab.json')
+#SAMPLE_HUMAN_VOCAB = os.path.join(HERE, 'data', 'sample_human_vocab.json')
+#SAMPLE_MACHINE_VOCAB = os.path.join(HERE, 'data', 'sample_machine_vocab.json')
+SAMPLE_HUMAN_VOCAB = os.path.join(HERE, 'data', 'human_vocab.json')
+SAMPLE_MACHINE_VOCAB = os.path.join(HERE, 'data', 'machine_vocab.json')
```

Please let me know if you need any other details.

**Comment by zafarali** [comment]
Steps being taken:

```
cd data
rm *_vocab.json
rm *.csv
python3 generate.py
cd ../weights
rm *.hdf5
cd ..
python3 run.py
python3 visualize.py -e "june 8, 2015" -w ./weights/NMT.48-0.01.hdf5  -hv ./data/human_vocab.json -mv ./data/machine_vocab.json
```

I am able to obtain a good visualization:

![image](https://user-images.githubusercontent.com/6295292/29132764-4722181e-7cff-11e7-8191-fcfe7c92f6fb.png)

**Comment by zafarali** [comment]
You mentioned you used `weights/NMT.07-0.02.hdf5`, this is the 7th epoch. Do you have any epoch higher than this? For example `NMT.48` or `NMT.49`?

**Comment by yeshPP** [comment]
Interesting. I changed run.py to be able to specify a weight checkpoint (see diff below), and I was training in increments of 5, 10 or 15 epochs to see when I can get good visualization from the model. "NMT.07-0.02.hdf5" was the result after the last set of 10 epochs. The loss and accuracy after that are below:

```
Epoch 10/10
100/100 [==============================] - 172s - loss: 0.0153 - acc: 0.9952 - all_acc: 0.0000e+00 - val_loss: 0.0166 - val_acc: 0.9948 - val_all_acc: 0.0000e+00
Model training complete. 
```

I will clear my weights folder, start from scratch and run it for 50 epochs and let you know if that fixes the attention map. Thanks again for looking into this. 

Changes to run.py: 
```
diff --git a/run.py b/run.py
index eb09973..59df573 100644
--- a/run.py
+++ b/run.py
@@ -6,6 +6,7 @@ import os
 import argparse

 from keras.callbacks import ModelCheckpoint
+from keras.models import load_model

 from models.NMT import simpleNMT
 from data.reader import Data, Vocabulary

@@ -51,6 +53,8 @@ def main(args):
                       decoder_units=256,
                       trainable=True,
                       return_probabilities=False)
+    if os.path.exists(args.weights):
+        model.load_weights(args.weights, by_name=True)

     model.summary()
     model.compile(optimizer='adam',
@@ -103,6 +107,10 @@ if __name__ == '__main__':
     named_args.add_argument('-b', '--batch-size', metavar='|',
                             help="""Location of validation data""",
                             required=False, default=32, type=int)
+
+    named_args.add_argument('-w', '--weights', metavar='|',
+                            help="""Previous Checkpoint weights""",
+                            required=False, default=0, type=str)
     args = parser.parse_args()
     print(args)
```

**Comment by zafarali** [comment]
I see, I'm going to close this issue but please do keep the communication open. I would like to see the results of your experiments!

Did you fork the repo so I can check in? I cannot find it online.

**Comment by yeshPP** [comment]
Just to close out the conversation, I was able to get the attention maps to look reasonable. The attention maps look much better with smaller networks (say 32 or 64 LSTM units instead of 256) as it forces the attention layer to learn.  While the year and day mostly make sense, the attention map for the month doesn't always seem to make sense, and in a lot of cases it seems to have an offset of 1. Even in your map above for "June" you would expect it to pay attention to "n" instead of "e" since the "n" differentiates it from "juL".

**Comment by zafarali** [comment]
Agreed! It's quite interesting and might be neat to investigate further. 

Do let me know what you find!
